exchange/bitswap/bitswap.go

```diff
@@ -308,154 +253,76 @@ func (bs *Bitswap) sendWantlistToProviders(ctx context.Context, entries []wantli
 			defer cancel()
 			providers := bs.network.FindProvidersAsync(child, k, maxProvidersPerRequest)
 			for prov := range providers {
-				sendToPeers <- prov
+				go func(p peer.ID) {
+					bs.network.ConnectTo(ctx, p)
+				}(prov)
 			}
 		}(e.Key)
 	}
 
-	go func() {
-		wg.Wait() // make sure all our children do finish.
-		close(sendToPeers)
-	}()
-
-	err := bs.sendWantlistToPeers(ctx, sendToPeers)
-	if err != nil {
-		log.Debugf("sendWantlistToPeers error: %s", err)
-	}
+	wg.Wait() // make sure all our children do finish.
```

Comment:
after looking at it, I think this is an untested assumption:

- https://github.com/ipfs/go-ipfs/blob/master/routing/dht/routing.go#L195
- https://github.com/ipfs/go-ipfs/blob/master/routing/dht/routing.go#L199-L203
- https://github.com/ipfs/go-ipfs/blob/master/routing/dht/routing.go#L259
- https://github.com/ipfs/go-ipfs/blob/master/routing/dht/routing.go#L214-L217
- https://github.com/ipfs/go-ipfs/blob/master/routing/dht/routing.go#L218-L221
- https://github.com/ipfs/go-ipfs/blob/master/routing/dht/routing.go#L232-L237
- https://github.com/ipfs/go-ipfs/blob/master/routing/dht/routing.go#L250-L254

any of these places could screw up, not respect contexts anymore and this breaks.


Now, i'm generally very much on the side of "if you think things are supposed to exit, WAIT for them, so that you find bugs asap" (like panic, basically). but i do think we should be testing our assumptions here. the dht tests are flakey and this is coupling the two things together.  :/  

we can keep it this way, and hope for the best. or try to test things and possibly waste time. \o/ decisions!